pupil-spa/src/app/services/azure-queue/azure-queue.service.ts

```diff
@@ -0,0 +1,59 @@
+import * as bluebird from 'bluebird';
+import { WindowRefService } from '../window-ref/window-ref.service';
+import { Injectable } from '@angular/core';
+import { TokenService } from '../token/token.service';
+
+/**
+ * Declaration of azure queue service
+ */
+@Injectable()
+export class AzureQueueService {
+  private window;
+  private queueService;
+
+  constructor(private tokenService: TokenService,
+              private windowRefService: WindowRefService) {
+    this.window = windowRefService.nativeWindow;
+  }
+
+  /**
+   * Initialise queue service for messaging
+   * @param {String} queue
+   * @param {String} tokenKey
+   * @returns {Promise}
+   */
+  private initQueueService(queue, tokenKey): void {
+    const token = this.tokenService.getToken(tokenKey);
+    if (!this.queueService) {
+      this.queueService = this.window.AzureStorage.Queue
+        .createQueueServiceWithSas(token.url.replace(queue, ''), token.token)
+        .withFilter(new this.window.AzureStorage.Queue.ExponentialRetryPolicyFilter()
+        );
+    }
+  }
+
+  /**
+   * Setup promisified queue service methods required for message creation
+   * @returns {Promise}
+   */
+  private initQueueMessaging(): void {
+      this.queueService.performRequest = bluebird.promisify(this.queueService.performRequest, this.queueService);
+      this.queueService.createMessage = bluebird.promisify(this.queueService.createMessage, this.queueService);
+  }
+
+  /**
+   * Add message to the queue
+   * @param {String} queue
+   * @param {String} tokenKey
+   * @param {Object} payload
+   * @returns {Promise.<void>}
+   */
+  public async addMessage(queue, tokenKey, payload): Promise<void> {
```

Comment:
I'm not sure this library signature is right.   I would prefer passing the token into `addMessage` rather than the `tokenKey`.  Reason being that passing in the `tokenKey` requires an understanding of our local storage system and increases dependencies.